## Airbnb DC Hosting Helper ##

## 5_modeling ##

### Summary ###

In this notebook, I will be exploring different modeling tecniques to see what the best fit is for my data. 

Import libraries and read in data

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn import preprocessing
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.svm import LinearSVC
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import cross_val_score
from sklearn.metrics import precision_score, recall_score, confusion_matrix, classification_report, accuracy_score, f1_score, ConfusionMatrixDisplay
from sklearn import metrics
from sklearn.metrics import roc_curve, auc, roc_auc_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier, GradientBoostingClassifier, ExtraTreesClassifier
from xgboost import XGBClassifier
from sklearn.naive_bayes import BernoulliNB

In [2]:
np.random.seed(100)

In [3]:
pd.set_option('display.max_columns', 300)

In [4]:
pd.set_option('display.max_rows', 300)

In [19]:
df = pd.read_csv('../data/cleaned_numerical_df.csv')

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3652 entries, 0 to 3651
Data columns (total 48 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Unnamed: 0                      3652 non-null   int64  
 1   id                              3652 non-null   int64  
 2   name                            3652 non-null   object 
 3   description                     3609 non-null   object 
 4   neighborhood_overview           2796 non-null   object 
 5   host_id                         3652 non-null   int64  
 6   host_about                      2538 non-null   object 
 7   host_response_time              3652 non-null   object 
 8   host_response_rate              3652 non-null   float64
 9   host_acceptance_rate            3652 non-null   float64
 10  host_is_superhost               3652 non-null   int64  
 11  host_has_profile_pic            3652 non-null   int64  
 12  host_identity_verified          36

Drop unnecessary columns for modeling.

In [21]:
df.drop(columns = ['Unnamed: 0', 'id', 'name', 'description', 'neighborhood_overview',
                   'host_id', 'host_about', 'latitude_x', 'longitude_x', 'amenities'
                  ], inplace=True)

Get dummies for categorical columns

In [22]:
df = pd.get_dummies(df,columns = ['host_response_time', 'neighbourhood_cleansed', 'room_type'], drop_first=True)

In [23]:
df.head()

host_response_rate  host_acceptance_rate  host_is_superhost  \
0                 0.8                  0.75                  0   
1                 1.0                  0.93                  0   
2                 1.0                  0.35                  0   
3                 1.0                  1.00                  0   
4                 1.0                  1.00                  0   

   host_has_profile_pic  host_identity_verified  accommodates  bathrooms  \
0                     1                       1             1        1.0   
1                     1                       1             2        1.0   
2                     1                       1             4        1.0   
3                     1                       1             4        1.0   
4                     1                       1             2        1.5   

   bedrooms  beds  price  minimum_nights  maximum_nights  \
0       1.0   1.0   55.0               2             365   
1       1.0   2.0   97.0               7             200   
2       1.0   2.0  185.0               2             180   
3       1.0   3.0  125.0               1             365   
4       1.0   1.0   61.0               1             365   

   minimum_nights_avg_ntm  maximum_nights_avg_ntm  availability_30  \
0                     2.0                   365.0                1   
1                     7.0                  1125.0                9   
2                     2.0                   180.0               17   
3                     1.0                  1125.0               12   
4                     1.0                  1125.0               19   

   availability_60  availability_90  availability_365  instant_bookable  \
0               31               61               336                 0   
1               20               50               140                 0   
2               47               76                76                 0   
3               42               72               347                 1   
4               49               79               354                 1   

   calculated_host_listings_count  historic site  museum  metro  music venue  \
0                               2              1       2      1            0   
1                              43              2       1      0            0   
2                               2             36      28      2           15   
3                               4              5       0      0            1   
4                               4              5       0      0            1   

   perfomring arts venue  college and university  food  nightlife spot  \
0                      3                      10    25               5   
1                      1                      40    18               4   
2                     46                      44    50              50   
3                      4                      17    48              27   
4                      4                      17    48              27   

   outdoors and recreation  government building  clothing store  popular  \
0                       28                   21               5        0   
1                       29                    6               0        1   
2                       45                   48              47        0   
3                       44                   24               2        0   
4                       44                   24               2        0   

   days_being_host  days_since_first_review  days_since_last_review  \
0             4610                     2576                     180   
1             4437                     1912                      16   
2             4346                     2223                      19   
3             4347                     3929                      23   
4             4347                     2121                      28   

   host_response_time_unknown  host_response_time_within a day  \
0                           0           

Set testing and training groups

In [24]:
X = df.drop(columns=['popular'])

y = df['popular']

X_train, X_test, y_train, y_test = train_test_split(X, y, stratify = y )

In [25]:
ss = StandardScaler()

X_train_sc = ss.fit_transform(X_train)

X_test_sc = ss.transform(X_test)

In [26]:
def pipe(model):
    #instantiate model
    model = model()
    #fit to scaled data
    model.fit(X_train_sc, y_train)
    
    #make predictions
    predictions = model.predict(X_test_sc)

    #print results
    print(f'{model} training score: {model.score(X_train_sc, y_train)}')
    print(f'{model} testing score: {model.score(X_test_sc, y_test)}')

    return

In [27]:
pipe(LogisticRegression)

LogisticRegression() training score: 0.8357064622124863
LogisticRegression() testing score: 0.8466593647316539


In [28]:
pipe(DecisionTreeClassifier)

DecisionTreeClassifier() training score: 1.0
DecisionTreeClassifier() testing score: 0.8138006571741512


In [29]:
pipe(GradientBoostingClassifier)

GradientBoostingClassifier() training score: 0.9087258123402702
GradientBoostingClassifier() testing score: 0.8630887185104053


In [30]:
pipe(AdaBoostClassifier)

AdaBoostClassifier() training score: 0.8678349762687112
AdaBoostClassifier() testing score: 0.8466593647316539


In [31]:
pipe(RandomForestClassifier)

RandomForestClassifier() training score: 1.0
RandomForestClassifier() testing score: 0.8532311062431545


In [32]:
pipe(SVC)

SVC() training score: 0.8587075575027382
SVC() testing score: 0.8510405257393209


In [47]:
pipe(BernoulliNB)

BernoulliNB() training score: 0.8079591091639284
BernoulliNB() testing score: 0.8324205914567361


In [34]:
pipe(KNeighborsClassifier)

KNeighborsClassifier() training score: 0.8532311062431545
KNeighborsClassifier() testing score: 0.7973713033953997


In [33]:
logreg = LogisticRegression()

logreg.fit(X_train_sc, y_train)

predictions = logreg.predict(X_test_sc)

print(f'training score: {logreg.score(X_train_sc, y_train)}')
print(f'testing score: {logreg.score(X_test_sc, y_test)}')

training score: 0.8357064622124863
testing score: 0.8466593647316539


In [44]:
sorted(set(zip(logreg.coef_[0], X.columns)), reverse=True)

[(1.1061461623865914, 'days_since_first_review'),
 (0.9224258202238735, 'host_is_superhost'),
 (0.5017011279325907, 'accommodates'),
 (0.2910542683249673, 'minimum_nights'),
 (0.28130804748358, 'host_acceptance_rate'),
 (0.18161039229625672, 'host_has_profile_pic'),
 (0.14774008691026486, 'neighbourhood_cleansed_West End, Foggy Bottom, GWU'),
 (0.14772158131632482, 'neighbourhood_cleansed_Brookland, Brentwood, Langdon'),
 (0.14268857522242176,
  'neighbourhood_cleansed_Woodridge, Fort Lincoln, Gateway'),
 (0.10095969126110649, 'maximum_nights'),
 (0.0955933455900523, 'availability_90'),
 (0.09546067200988377, 'host_identity_verified'),
 (0.09519088470296674, 'neighbourhood_cleansed_Eastland Gardens, Kenilworth'),
 (0.07834970059334948, 'host_response_rate'),
 (0.0768194492434986,
  'neighbourhood_cleansed_Mayfair, Hillbrook, Mahaning Heights'),
 (0.07314851560072703,
  'neighbourhood_cleansed_Friendship Heights, American University Park, Tenleytown'),
 (0.0713956343510424, 'college and

As days since first review increases by 1, the log-odds of someone being a popular listing increases by 1.11.

In [45]:
sorted(set(zip(np.exp(logreg.coef_[0]), X.columns ) ), reverse=True)


[(3.0226869742470863, 'days_since_first_review'),
 (2.515384866375245, 'host_is_superhost'),
 (1.651528343424749, 'accommodates'),
 (1.337837184163135, 'minimum_nights'),
 (1.324861661382376, 'host_acceptance_rate'),
 (1.199146905983806, 'host_has_profile_pic'),
 (1.1592115629454387, 'neighbourhood_cleansed_West End, Foggy Bottom, GWU'),
 (1.159190111245453, 'neighbourhood_cleansed_Brookland, Brentwood, Langdon'),
 (1.1533705575608024,
  'neighbourhood_cleansed_Woodridge, Fort Lincoln, Gateway'),
 (1.1062320500458507, 'maximum_nights'),
 (1.1003115264690373, 'availability_90'),
 (1.1001655538830868, 'host_identity_verified'),
 (1.0998687832154215, 'neighbourhood_cleansed_Eastland Gardens, Kenilworth'),
 (1.081500794018528, 'host_response_rate'),
 (1.0798470915815115,
  'neighbourhood_cleansed_Mayfair, Hillbrook, Mahaning Heights'),
 (1.0758903115458327,
  'neighbourhood_cleansed_Friendship Heights, American University Park, Tenleytown'),
 (1.074006055513004, 'college and university'),


As days since first review increases by 1, the listing is 3.02 times more likely to be popular.

* Reminder that this is an increase of scaled units. We just need to make this caviot for interpreting results. reminder of what standard scalar does (mean of 0 std of 1?)

In [60]:
test = pd.DataFrame(y_test)

test['preds'] = predictions

test['accuracy'] = test['popular'] - test['preds']

In [63]:
new = test[~(test['accuracy']==0)]

new.shape

(140, 3)

In [64]:
new

popular  preds  accuracy
956         1      0         1
1180        0      1        -1
1132        1      0         1
955         0      1        -1
867         1      0         1
350         0      1        -1
852         0      1        -1
655         1      0         1
498         1      0         1
2327        1      0         1
14          0      1        -1
1403        1      0         1
1871        1      0         1
1096        1      0         1
627         1      0         1
398         0      1        -1
2787        1      0         1
1047        0      1        -1
1969        1      0         1
1261        1      0         1
1000        1      0         1
606         0      1        -1
2181        1      0         1
1186        1      0         1
2864        1      0         1
709         1      0         1
420         0      1        -1
1671        1      0         1
1343        1      0         1
211         1      0         1
75          0      1        -1
1530        1      0         1
399         0      1        -1
190         1      0         1
307         1      0         1
339         0      1        -1
327         0      1        -1
1273        1      0         1
1130        1      0         1
1708        1      0         1
1160        1      0         1
58          0      1        -1
32          0      1        -1
1028        1      0         1
25          0      1        -1
213         0      1        -1
663         1      0         1
2132        1      0         1
488         1      0         1
48          0      1        -1
123         1      0         1
834         1      0         1
1324        0      1        -1
1253        1      0         1
1537        0      1        -1
2577        1      0         1
1778        1      0         1
1768        1      0         1
1609        1      0         1
1625        1      0         1
459         1      0         1
1158        0      1        -1
1295        1      0         1
409         0      1        -1
918         1      0         1
2930        1      0         1
1772        1      0         1
1276        0      1        -1
2228        1      0         1
1464        1      0         1
1283        1      0         1
1447        0      1        -1
1457        0      1        -1
2397        1      0         1
81          1      0         1
539         0      1        -1
1409        1      0         1
1237        0      1        -1
2512        1      0         1
989         0      1        -1
1044        1      0         1
1810        1      0         1
250         0      1        -1
556         1      0         1
1756        1      0         1
1208        1      0         1
1567        1      0         1
533         1      0         1
89          1      0         1
951         1      0         1
1653        1      0         1
235         1      0         1
233         0      1        -1
287         1      0         1
1026        1      0         1
423         1      0         1
175         1      0         1
2209        1      0         1
1798        0      1        -1
2248        1      0         1
1947        1      0         1
1234        1      0         1
227         1      0         1
701         1      0         1
1115        1      0         1
2061        1      0         1
625         0      1        -1
1037        1      0         1
608         0      1        -1
1525        1      0         1
1570        1      0         1
266         0      1        -1
1087        0      1        -1
1628        1      0         1
1030        1      0         1
387         0      1        -1
1356        0      1        -1
1845        0      1        -1
1813        1      0         1
3187        0      1        -1
200         0      1        -1
283         0      1        -1
428         0      1        -1
1793        1      0         1
1603        1      0         1
976         0      1        -1
524         1      0         1
1795        1      0         1
21     